File: app/src/main/java/fi/kaupunkifillarit/maps/RackMarkerOptions.java
package fi.kaupunkifillarit.maps;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Typeface;

import fi.kaupunkifillarit.R;
import fi.kaupunkifillarit.model.Rack;

public class RackMarkerOptions {
    private final Maps.MarkerOptionsWrapper status;

    public RackMarkerOptions(Rack rack, Resources res, Maps.MapWrapper map) {
        this.status = map.createMarkerOptions()
                .icon(RackMarkerOptions.getMarkerBitmap(
                        rack.bikes < 2,
                        rack.bikes + "/" + rack.slots,
                        res))
                .flat(false)
                .position(rack.latitude, rack.longitude)
                .anchor(0.5f, 1f);
    }

    private static Bitmap getMarkerBitmap(boolean empty, String text, Resources res) {
        Typeface regular = Typeface.createFromAsset(res.getAssets(), "fonts/Montserrat-Regular.ttf");
        Paint paint = new Paint();
        paint.setTypeface(regular);
        paint.setTextSize(res.getDimensionPixelSize(R.dimen.indicator_text_size));
        paint.setColor(Color.BLACK);
        paint.setTextAlign(Paint.Align.CENTER);
        paint.setAntiAlias(true);
        Bitmap marker = BitmapFactory.decodeResource(res, empty ? R.drawable.rack_marker_empty : R.drawable.rack_marker);
        Bitmap bitmap = Bitmap.createBitmap(marker.getWidth(), marker.getHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        canvas.drawBitmap(marker, 0, 0, null);
        canvas.drawText(text, (float) canvas.getWidth() / 2f, (float) canvas.getHeight() / 2f, paint);

        return bitmap;
    }

    public RackMarker makeMarker(Maps.MapWrapper map) {
        return new RackMarker(map.addMarker(status));
    }
}

File: app/src/main/kotlin/fi/kaupunkifillarit/maps/RackMarkerOptions.kt
package fi.kaupunkifillarit.maps

import android.content.res.Resources
import android.graphics.*
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import fi.kaupunkifillarit.R
import fi.kaupunkifillarit.model.Rack

class RackMarkerOptions {
    private val status: Maps.MarkerOptionsWrapper<*>

    constructor(rack: Rack, res: Resources, map: Maps.MapWrapper<Maps.MarkerWrapper<*>, Maps.MarkerOptionsWrapper<*>>) {
        this.status = map.createMarkerOptions().icon(getMarkerBitmap(
                rack.bikes < 2,
                "${rack.bikes}",
                res)).flat(false).position(rack.latitude, rack.longitude).anchor(0.5f, 1f)
    }

    private fun getMarkerBitmap(isEmpty: Boolean, text: String, res: Resources): Bitmap {
        val regular = Typeface.createFromAsset(res.assets, "fonts/Montserrat-Regular.ttf")
        val paint = Paint()
        paint.typeface = regular
        paint.textSize = res.getDimensionPixelSize(R.dimen.indicator_text_size).toFloat()
        paint.color = Color.BLACK
        paint.textAlign = Paint.Align.CENTER
        paint.isAntiAlias = true
        val marker = BitmapFactory.decodeResource(res, if (isEmpty) R.drawable.rack_marker_empty else R.drawable.rack_marker)
        val bitmap = Bitmap.createBitmap(marker.width, marker.height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        canvas.drawBitmap(marker, 0f, 0f, null)
        canvas.drawText(text, canvas.width.toFloat() / 2f, canvas.height.toFloat() / 2f, paint)

        return bitmap
    }

    fun makeMarker(map: Maps.MapWrapper<Maps.MarkerWrapper<*>, Maps.MarkerOptionsWrapper<*>>): RackMarker {
        return RackMarker(map.addMarker(status))
    }
}

object RackMarkerBitmap {
    fun bitmap(isEmpty: Boolean, text: String, res: Resources): Bitmap {
        val regular = Typeface.createFromAsset(res.assets, "fonts/Montserrat-Regular.ttf")
        val paint = Paint()
        paint.typeface = regular
        paint.textSize = res.getDimensionPixelSize(R.dimen.indicator_text_size).toFloat()
        paint.color = Color.BLACK
        paint.textAlign = Paint.Align.CENTER
        paint.isAntiAlias = true
        val marker = BitmapFactory.decodeResource(res, if (isEmpty) R.drawable.rack_marker_empty else R.drawable.rack_marker)
        val bitmap = Bitmap.createBitmap(marker.width, marker.height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        canvas.drawBitmap(marker, 0f, 0f, null)
        canvas.drawText(text, canvas.width.toFloat() / 2f, canvas.height.toFloat() / 2f, paint)

        return bitmap
    }
}

fun MarkerOptions.rackMarker(rack: Rack, res: Resources) =
        this.icon(BitmapDescriptorFactory.fromBitmap(RackMarkerBitmap.bitmap(rack.bikes < 2, "${rack.bikes}", res)))
                .flat(true)
                .position(LatLng(rack.latitude, rack.longitude))
                .anchor(0.5f, 1f)

